util/common/src/main/java/org/hibernate/search/util/common/jar/impl/CodeSource.java

```diff
@@ -169,11 +193,22 @@ private String extractedJarNestedPath(URI jarUri) throws IOException {
 		else {
 			int afterPathSeparatorIndex = pathSeparatorIndex + JAR_URI_PATH_SEPARATOR.length();
 			int secondPathSeparatorIndex = spec.indexOf( JAR_URI_PATH_SEPARATOR, afterPathSeparatorIndex );
-			if ( 0 <= secondPathSeparatorIndex && secondPathSeparatorIndex + JAR_URI_PATH_SEPARATOR.length() < spec.length() ) {
-				// TODO HSEARCH-4744 support reading the content of nested JARs
-				throw log.cannotOpenNestedJar( jarUri );
+			while ( 0 <= secondPathSeparatorIndex ) {
+				Path nestedPathInFileSystem = classesPathInFileSystem.resolve(
+						spec.substring( afterPathSeparatorIndex, secondPathSeparatorIndex )
+				);
+				if ( Files.isRegularFile( nestedPathInFileSystem ) ) {
+					nonDefaultFileSystem = NESTED_JAR_FILESYSTEM_CREATOR.apply( nestedPathInFileSystem, jarUri );
```

Comment:
This is a loop. If you're overwriting the previous filesystem, it probably won't ever be closed.

I'd suggest to keep a list of filesystems to close in `CodeSource`?